### PR TITLE
[SHEPHERD]  added the skeletons for the Alliance and Goal classes, and Utils files

### PR DESCRIPTION
**Pull request by FIREdog5**
Added the skeleton methods for the Goal class, including `__init__`, `score`, `calc_multiplier`, and `apply_powerup`, as well as all the instance variables.

Added the skeleton methods for Alliance class, including `__init__`, and `change_score`, as well as the instance variables.

Added the utils file, which holds several important constants.

### CODE REVIEW

File: shepherd/Alliance.py [hunk 1]
```diff
@@ -1,37 +1,42 @@
...
-    '''
+    """This is the Alliance class, which holds the state values used to track
+       the scores and states of the alliances
+            name - an Enum 'GOLD' or 'BLUE' representing the Alliance
+            team_1_name - string representing name of first team
+            team_2_name - string representing name of second team
+            team_1_number - Integer representing team number of first team
+            team_2_number - Integer representing team number of second team
+            score - Integer tracking the score of the Alliance
+            alliance_multiplier - Integer tracking the value of the alliance's
```
Comment (brandonxxlee) [inline shepherd/Alliance.py]: Technically a float

File: shepherd/Alliance.py [hunk 2]
```diff
@@ -1,37 +1,42 @@
...
+    """
 
-    def __init__(self, team1, team2):
-        self.team_one = team1
-        self.team_two = team2
-        self.score = 0
-        self.large_penalties = 0
-        self.small_penalties = 0
-        self.multiplier = 1
+    def __init__(self, name, team_1_name, team_2_name, team_1_number,
```
Comment (brandonxxlee) [inline shepherd/Alliance.py]: Score should just be set to 0, does not need to an argument into the constructor

File: shepherd/Alliance.py [hunk 3]
```diff
@@ -1,37 +1,42 @@
...
+                 team_2_number, score):
+        self.name = name
+        self.team_1_name = team_1_name
+        self.team_2_name = team_2_name
+        self.team_1_number = team_1_number
+        self.team_2_number = team_2_number
+        self.score = score
+        self.alliance_multiplier = 1
+        self.two_x_cooldown = Timer()
```
Comment (brandonxxlee) [inline shepherd/Alliance.py]: Timer constructor should take in some arguments. See shepherd/master for Timer.py

File: shepherd/Alliance.py [hunk 4]
```diff
@@ -1,37 +1,42 @@
...
-        Adds an additional small penalty to this alliance
-        '''
-        return
-
-    def increase_score(self, amount): # pylint: disable=unused-argument
-        '''
-        Increases this alliance's score by amount
-        '''
-        return
+    def change_score(ammount):
```
Comment (brandonxxlee) [inline shepherd/Alliance.py]: typo: ammount

File: shepherd/Alliance.py [hunk 5]
```diff
@@ -1,37 +1,42 @@
...
+        """ changes score of this alliance by Amount,
+            Ammount can be negative or positive.
+            make sure you update the scoreboad too after using this function
+        """
+        self.score += ammount
```
Comment (brandonxxlee) [inline shepherd/Alliance.py]: type: ammount

File: shepherd/Alliance.py [hunk 6]
```diff
@@ -1,37 +1,42 @@
...
-        return
-
-    def increase_score(self, amount): # pylint: disable=unused-argument
-        '''
-        Increases this alliance's score by amount
-        '''
-        return
+    def change_score(ammount):
+        """ changes score of this alliance by Amount,
+            Ammount can be negative or positive.
```
Comment (brandonxxlee) [inline shepherd/Alliance.py]: typo: ammount

File: shepherd/Goal.py [hunk 7]
```diff
@@ -1,41 +1,63 @@
-"""
+import Utils.py
```
Comment (brandonxxlee) [inline shepherd/Goal.py]: include newline after this. We might want to from Utils.py import * eventually too.

File: shepherd/Goal.py [hunk 8]
```diff
@@ -1,41 +1,63 @@
...
-    BiddingMulitplierValue
-    BiddingTimeStartVAlue
-    BiddingAdditionaTimeValue
-    ScoreMultiplierValues
-    GoalMultiplierValues
-    GoalMulitpliercooldowns
-
-"""
+    """This class holds the state values for the goal.
+            value - An integer representing the point value of the Goal
```
Comment (brandonxxlee) [inline shepherd/Goal.py]: Clarify this more, we'll probably forget what this means

File: shepherd/Goal.py [hunk 9]
```diff
@@ -1,41 +1,63 @@
...
+            current_bid - An integer representing the value of the current bid
+                          on this goal
+            current_bid_team - An Alliance object representing the team that
+                               has submited the most recent bid
+            previous_bid - An integer representing the value of the previous
+                           bid
+            previous_bid_team - An alliance object representing the team that
+                                made the previous bid
+            bid_timer - A timer tracking the time left to bid
+            gold_two_x_timer - A timer tracking how long the 2x powerup applied
```
Comment (brandonxxlee) [inline shepherd/Goal.py]: Capitalize class object names (e.g. Timer)

File: shepherd/Goal.py [hunk 10]
```diff
@@ -1,41 +1,63 @@
...
+    """
+    bid_increase_constant = Utils.BID_INCREASE_CONSTANT
+    def __init__(self, value):
+        self.value = value
+        self.owner = None
+        self.current_bid = 0
+        self.current_bid_team = None
+        self.previous_bid = 0
+        self.previous_bid_team = None
+        self.bid_timer = Timer()
```
Comment (brandonxxlee) [inline shepherd/Goal.py]: See above comment about including args into Timer

File: shepherd/Goal.py [hunk 11]
```diff
@@ -1,41 +1,63 @@
...
+        self.gold_zero_x_timer = Timer()
+        self.blue_two_x_timer = Timer()
+        self.blue_zero_x_timer = Timer()
+
+    def score(alliance):
+        """ modifies the owner alliance's score based on the value of the goal,
+            the current multipliers,and the alliances multiplier, if the
+            scorring Alliance is the owning alliance.
+                alliance - Alliance object representing the scoring team
+        """
```
Comment (brandonxxlee) [inline shepherd/Goal.py]: I think you have to include pass in empty functions

File: shepherd/Utils.py [hunk 12]
```diff
@@ -0,0 +1,7 @@
+
+# Values for the cooldowns
```
Comment (brandonxxlee) [inline shepherd/Utils.py]: Wrap this into a class (e.g. CONSTANTS or SHEPHERD_CONSTANTS). when calling from classes use from shepherd import *, and use CONSTANTS.CONSTANT_NAME

**Review by brandonxxlee — CHANGES_REQUESTED** [review]
Left some feedback

File: shepherd/Goal.py [hunk 13]
```diff
@@ -1,41 +1,71 @@
...
+    def __init__(self, name, value):
+        self.name = name
+        self.value = value
+        self.owner = None
+        self.current_bid = 0
+        self.current_bid_team = None
+        self.previous_bid = 0
+        self.previous_bid_team = None
+        self.bid_timer = Timer('bid', name)
+        self.gold_two_x_timer = Timer('duration', name)
```
Comment (brandonxxlee) [inline shepherd/Goal.py]: You should make all of these (e.g. 'duration', 'cooldown', etc) also a class in Utils.py

File: shepherd/Utils.py [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+from enum import Enum, unique
+
+@unique # pylint: disable=invalid-name
+class CONSTANTS(Enum):
+    TWO_X_COOLDOWN = 1
```
Comment (brandonxxlee) [inline shepherd/Utils.py]: I think these need to map to different values. Not exactly sure how enums work in python to be honest.

File: shepherd/Alliance.py [hunk 15]
```diff
@@ -1,37 +1,49 @@
...
+            make sure you update the scoreboad too after using this function
+        """
+        self.score += amount
 
-    def increase_score(self, amount): # pylint: disable=unused-argument
-        '''
-        Increases this alliance's score by amount
-        '''
-        return
+    def i_hate_pylint(self):
```
Comment (brandonxxlee) [inline shepherd/Alliance.py]: Just disable the error in .pylintrc instead of this

**Review by brandonxxlee — CHANGES_REQUESTED** [review]
A couple more changes.

**Review by brandonxxlee — DISMISSED** [review]
lgtm

File: shepherd/Goal.py [hunk 16]
```diff
@@ -1,41 +1,71 @@
...
+        self.owner = None
+        self.current_bid = 0
+        self.current_bid_team = None
+        self.previous_bid = 0
+        self.previous_bid_team = None
+        self.bid_timer = Timer(TIMER_TYPES.BID, name)
+        self.gold_two_x_timer = Timer(TIMER_TYPES.DURATION, name)
+        self.gold_zero_x_timer = Timer(TIMER_TYPES.DURATION, name)
+        self.blue_two_x_timer = Timer(TIMER_TYPES.DURATION, name)
+        self.blue_zero_x_timer = Timer(TIMER_TYPES.DURATION, name)
```
Comment (JHGZhou) [inline shepherd/Goal.py]: Do these timers need the name passed in here? Where are we making use of these timers except for when we are checking the current multiplier?

File: shepherd/Goal.py [hunk 17]
```diff
@@ -1,41 +1,71 @@
...
+            return an Integer signifying what the result of the call was.
+                1 - success
+                0 - incorrect solution
+                -1 - cannot be applied to this goal
+                -2 - cannot steal your own goal
+                -3 - other error
+            effect - enum representing an effect to be applied to the goal
+            alliance - the alliance trying to apply the effect to the goal
+        """
+        pass
```
Comment (JHGZhou) [inline shepherd/Goal.py:70]: When will incorrect solution ever be returned? Is the solution being checked here? If so, then there needs to be another argument for the solution to come in.

File: shepherd/Timer.py [hunk 18]
```diff
@@ -4,6 +4,16 @@ class Timer:
...
+        """
+        timer_type - a Enum representing the type of timer that this is:
+                        TIMER_TYPES.BID - represents a bid timer
+                        TIMER_TYPES.POWERUP - represents a powerup timer
+                        TIMER_TYPES.COOLDOWN - represents a cooldown timer
+                        TIMER_TYPES.CODE_COOLDOWN - represents a cooldown timer
+                                                    for a code
+                        TIMER_TYPES.DURATION - represents a duration timer for
+                                               a powerup
+        """
```
Comment (JHGZhou) [inline shepherd/Timer.py:19]: What is the difference between the POWERUP AND DURATION timers?

File: shepherd/Goal.py (hunk 17 above)
Comment (FIREdog5) [reply shepherd/Goal.py:70]: No, this part will not check the solution.

**Review by JHGZhou — CHANGES_REQUESTED** [review]
Mostly good, just small things for clarification in the docstrings

**Review by JHGZhou — APPROVED** [review]
Looks good

**Pull request merged**